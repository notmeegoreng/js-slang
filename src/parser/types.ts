import type { Program } from 'estree'

import type { Context, Node, Chapter, Variant, SourceError } from '../types'

export type { ParserOptions as BabelOptions } from '@babel/parser'
export type { Options as AcornOptions } from 'acorn'

export interface Parser<TOptions> {
  parse(
    programStr: string,
    context: Context,
    options?: Partial<TOptions>,
    throwOnError?: boolean
  ): Program | null
  validate(ast: Program, context: Context, throwOnError?: boolean): boolean
}
export interface Rule<T extends Node> {
  /**
   * Name of the rule
   */
  name: string

  /**
   * Test snippets to test the behaviour of the rule. Providing no test snippets
   * means that this rule will not be tested when running unit tests.\
   * First element of the tuple is the code to test. Set the second element to `undefined`
   * if the snippet should not throw an error. Otherwise set it to the `explain()` value
   * of the error.
   */
  testSnippets?: [code: string, expected: string | undefined][]

  /**
   * Disable this rule for this chapter (inclusive) and above
   */
  disableFromChapter?: Chapter
  disableForVariants?: Variant[]
  checkers: {
    [name: string]: (node: T, ancestors: Node[]) => SourceError[]
  }
}
