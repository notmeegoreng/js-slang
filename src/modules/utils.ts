import * as _ from 'lodash'

import { RecursivePartial } from '../types'
import { ImportOptions } from './moduleTypes'
import { defaultAnalysisOptions } from './preprocessor/analyzer'

const exportDefaultStr = 'export default'
export function removeExportDefault(text: string) {
  if (text.startsWith(exportDefaultStr)) {
    text = text.substring(exportDefaultStr.length).trim()
  }

  return text
}

export function mergeImportOptions(src?: RecursivePartial<ImportOptions>): ImportOptions {
  const baseOptions = _.cloneDeep(defaultAnalysisOptions)
  return _.merge(baseOptions, src as any)
}

export const isSourceModule = (path: string) => !path.startsWith('.') && !path.startsWith('/')
