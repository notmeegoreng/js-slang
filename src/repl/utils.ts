import { Option } from '@commander-js/extra-typings'
import { Chapter, Variant } from '../types'

export function chapterParser(str: string): Chapter {
  let foundChapter: string | undefined

  if (/^[0-9]$/.test(str)) {
    // Chapter is fully numeric
    const value = parseInt(str)
    foundChapter = Object.keys(Chapter).find(chapterName => Chapter[chapterName] === value)
  } else {
    foundChapter = str
  }

  if (foundChapter === undefined) {
    throw new Error(`Invalid chapter value: ${str}`)
  }

  return Chapter[foundChapter]
}

export const chapterOption = new Option('--chapter <chapter>')
  .default(Chapter.SOURCE_4)
  .argParser(chapterParser)

export const variantOption = new Option('--variant <variant>')
  .default(Variant.DEFAULT)
  .choices(Object.values(Variant))
