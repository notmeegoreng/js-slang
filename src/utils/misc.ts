import { RuntimeSourceError } from '../errors/runtimeSourceError'

export class PromiseTimeoutError extends RuntimeSourceError {}

export const timeoutPromise = <T>(promise: Promise<T>, timeout: number) =>
  new Promise<T>((resolve, reject) => {
    const timeoutid = setTimeout(() => reject(new PromiseTimeoutError()), timeout)

    promise
      .then(res => {
        clearTimeout(timeoutid)
        resolve(res)
      })
      .catch(e => {
        clearTimeout(timeoutid)
        reject(e)
      })
  })

/**
 * Run the mapping function over the items, filtering out any items that
 * that the mapping function returned `undefined` for
 */
export function mapAndFilter<T, U>(items: T[], mapper: (input: T) => U | undefined) {
  return items.reduce((res, item) => {
    const newItem = mapper(item)
    if (newItem !== undefined) return [...res, newItem]
    return res
  }, [] as U[])
}

export function mapObject<T extends Record<any, any>, U>(
  obj: T,
  mapper: (key: keyof T, value: T[keyof T]) => U
): Record<keyof T, U> {
  return objectEntries(obj).reduce(
    (res, [key, value]) => ({
      ...res,
      [key]: mapper(key, value)
    }),
    {} as Record<keyof T, U>
  )
}

/**
 * Type safe `Object.keys`
 */
export function objectKeys<T extends string | number | symbol>(obj: Record<T, any>): T[] {
  return Object.keys(obj) as T[]
}

/**
 * Type safe `Object.values`
 */
export function objectValues<T>(obj: Record<any, T>) {
  return Object.values(obj) as T[]
}

type DeconstructRecord<T extends Record<any, any>> = {
  [K in keyof T]: [K, T[K]]
}

type EntriesOfRecord<T extends Record<any, any>> = DeconstructRecord<T>[keyof T]

export function objectEntries<T extends Record<any, any>>(obj: T): EntriesOfRecord<T>[] {
  return Object.entries(obj)
}
