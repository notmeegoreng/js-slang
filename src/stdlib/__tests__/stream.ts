import { Chapter, Variant } from '../../types'
import { expectParsedError, expectResult, expectResultsToEqual } from '../../utils/testing'

expectResultsToEqual(
  [
    ['empty stream is null', 'stream();', null],
    [
      'append',
      `
      equal(stream_to_list(stream_append(stream("string", 123), stream(456, null, undefined)))
      , list("string", 123, 456, null, undefined));
    `,
      true
    ],
    [
      'build_stream',
      'equal(stream_to_list(build_stream(x => x * x, 5)), list(0, 1, 4, 9, 16));',
      true
    ],
    ['enum_stream', 'equal(stream_to_list(enum_stream(1, 5)), list(1, 2, 3, 4, 5));', true],
    [
      'enum_stream with floats',
      'equal(stream_to_list(enum_stream(1.5, 5)), list(1.5, 2.5, 3.5, 4.5));',
      true
    ],
    [
      'filter',
      `
      equal(
        stream_to_list(
          stream_filter(x => x <= 4, stream(2, 10, 1000, 1, 3, 100, 4, 5, 2, 1000))
        )
      , list(2, 1, 3, 4, 2));
    `,
      true
    ],
    [
      'for_each',
      `  
      let sum = 0;
      stream_for_each(x => {
        sum = sum + x;
      }, stream(1, 2, 3));
      sum;
    `,
      6
    ],
    [
      'map',
      'equal(stream_to_list(stream_map(x => 2 * x, stream(12, 11, 3))), list(24, 22, 6));',
      true
    ],
    [
      'member',
      `
      equal(
        stream_to_list(stream_member("string", stream(1, 2, 3, "string", 123, 456, null, undefined))),
        list("string", 123, 456, null, undefined));
    `,
      true
    ],
    ['remove', 'stream_remove(1, stream(1));', null],
    ['remove not found', 'stream_to_list(stream_remove(2, stream(1)));', [1, null]],
    [
      'remove_all',
      `
      equal(stream_to_list(stream_remove_all(1, stream(1, 2, 3, 4, 1, 1, "1", 5, 1, 1, 6))),
        list(2, 3, 4, "1", 5, 6));
    `,
      true
    ],
    [
      'remove_all not found',
      'equal(stream_to_list(stream_remove_all(1, stream(2, 3, "1"))), list(2, 3, "1"));',
      true
    ],
    [
      'reverse',
      `
      equal(stream_to_list(
        stream_reverse(
          stream("string", null, undefined, null, 123))),
      list(123, null, undefined, null, "string"));
    `,
      true
    ],
    ['stream_ref', 'stream_ref(stream(1, 2, 3, "4", 4), 4);', 4],
    ['stream_tail works', 'head(stream_tail(stream(1, 2)));', 2],
    [
      'stream is properly created',
      `
      const s = stream(true, false, undefined, 1, x=>x, null, -123, head);
      const result = [];
      stream_for_each(item => {result[array_length(result)] = item;}, s);
      stream_ref(s,4)(22) === 22 && stream_ref(s,7)(pair('', '1')) === '1' && result;
    `,
      false
    ],
    ['stream_to_list works for null', 'stream_to_list(null);', null]
  ],
  Chapter.SOURCE_3
)

test('stream_tail is lazy', () => {
  return expectResult('stream_tail(integers_from(0));', Chapter.SOURCE_3).toMatchInlineSnapshot(`
Array [
  1,
  [Function],
]
`)
})

describe('infinite stream is infinite', () => {
  // TODO: Make this more modular
  const code = 'stream_length(integers_from(0));'

  test('cse-machine', () => {
    return expectParsedError(code, {
      chapter: Chapter.SOURCE_3,
      variant: Variant.EXPLICIT_CONTROL
    }).toMatch(/(Maximum call stack size exceeded){1,2}/)
  }, 15000)

  test('native', () => {
    return expectParsedError(code, Chapter.SOURCE_3).toEqual(
      'Line 1: The error may have arisen from forcing the infinite stream: function integers_from.'
    )
  }, 15000)
})

test('stream_to_list works', () => {
  return expectResult(`stream_to_list(stream(1, true, 3, 4.4, [1, 2]));`, Chapter.SOURCE_3)
    .toMatchInlineSnapshot(`
Array [
  1,
  Array [
    true,
    Array [
      3,
      Array [
        4.4,
        Array [
          Array [
            1,
            2,
          ],
          null,
        ],
      ],
    ],
  ],
]
`)
})
