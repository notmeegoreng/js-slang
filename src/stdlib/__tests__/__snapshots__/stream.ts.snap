// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`for_each: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "let sum = 0;
stream_for_each(x => {
  sum = sum + x;
}, stream(1, 2, 3));
sum;",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 6,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`list_ref: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "stream_ref(stream(1, 2, 3, \\"4\\", 4), 4);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 4,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`primitive stream functions empty stream is null: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "stream();",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": null,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`primitive stream functions stream is properly created: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const s = stream(true, false, undefined, 1, x=>x, null, -123, head);
const result = [];
stream_for_each(item => {result[array_length(result)] = item;}, s);
stream_ref(s,4)(22) === 22 && stream_ref(s,7)(pair('', '1')) === '1' && result;",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": false,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`primitive stream functions stream_tail is lazy: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "stream_tail(integers_from(0));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "native:[1, () => integers_from(n + 1)]
interpreted:undefined",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`primitive stream functions stream_tail works: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "head(stream_tail(stream(1, 2)));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`primitive stream functions stream_to_list works for null: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "stream_to_list(null);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": null,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`primitive stream functions stream_to_list works: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "stream_to_list(stream(1, true, 3, 4.4, [1, 2]));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": Array [
    1,
    Array [
      true,
      Array [
        3,
        Array [
          4.4,
          Array [
            Array [
              1,
              2,
            ],
            null,
          ],
        ],
      ],
    ],
  ],
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`remove: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "stream_remove(1, stream(1));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": null,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;
